### PR TITLE
Add dmi_resize_png using the image crate

### PR DESCRIPTION
**Pull request by ShadowLarkens**
Please be gentle this is literally the first time I've ever touched rust ;w;
I am probably going to be absolutely helpless if this needs any optimization or tweaks to file writing

### CODE REVIEW

**Comment by optimumtact** [comment]
Thanks for the PR, and welcome aboard 👍

File: src/dmi.rs [hunk 1]
```diff
@@ -63,3 +67,13 @@ fn create_png(path: &str, width: &str, height: &str, data: &str) -> Result<()> {
     let mut writer = encoder.write_header()?;
     Ok(writer.write_image_data(&result)?)
 }
+
+fn resize_png(path: &str, width: &str, height: &str) -> Result<()> {
+    let width = u32::from_str_radix(width, 10)?;
+    let height = u32::from_str_radix(height, 10)?;
+
+    let img = image::open(path).unwrap();
```
Comment (SpaceManiac) [inline src/dmi.rs:75]: I would much rather return an error using `?` rather than panicking if loading or saving fails

**Comment by SpaceManiac** [comment]
Have you done any tests of how fast this is compared to http://www.byond.com/docs/ref/info.html#/icon/proc/Scale ?

**Comment by AnturK** [comment]
What's the use case here ?

**Comment by ShadowLarkens** [comment]
@AnturK @SpaceManiac The point of this is that Scale() has this little caveat: `Scale() automatically performs antialiasing to avoid unwanted artifacts.`  I have seen examples of this antialiasing *not* occuring, but... I have no idea why. There doesn't appear to be any consistent way to convince BYOND "Hey, don't just randomly apply bilinear antialiasing to this pixel art that I'm trying to upscale."

File: src/dmi.rs (hunk 1 above)
Comment (ShadowLarkens) [reply src/dmi.rs:75]: Doooes that just mean add a question mark to the end of this line? Could you make a code suggestion? I'm really unfamiliar with rust's syntax and error handling.

**Comment by ShadowLarkens** [comment]
Like, if anyone knows what the rhyme or reason to Scale() occasionally just deciding to apply bilinear antialiasing and how to prevent it from doing that, I'd be happy to implement my icon upscaling with native BYOND procs. This idea came up while working with the spritesheet asset datum and needing to upscale a spritesheet for a UI.

**Comment by ZeWaka** [comment]
> Like, if anyone knows what the rhyme or reason to Scale() occasionally just deciding to apply bilinear antialiasing and how to prevent it from doing that, I'd be happy to implement my icon upscaling with native BYOND procs.

You probably don't have `appearance_flags |= PIXEL_SCALE` set if you're performing this operation on an atom. However, that var doesn't seem to exist for /icon or /image, which could be your usecase.

**Comment by ShadowLarkens** [comment]
Yeah, it's direct /icon modification that matters.

File: src/dmi.rs (hunk 1 above)
Comment (ShadowLarkens) [inline src/dmi.rs:75]:
```suggestion
    let img = image::open(path)?;
```

File: src/dmi.rs [hunk 2]
```diff
@@ -63,3 +67,13 @@ fn create_png(path: &str, width: &str, height: &str, data: &str) -> Result<()> {
...
+    let newimg = img.resize(width, height, image::imageops::Nearest);
+
+    Ok(newimg.save_with_format(&path, image::ImageFormat::Png).unwrap())
```
Comment (ShadowLarkens) [inline src/dmi.rs:78]:
```suggestion
    Ok(newimg.save_with_format(&path, image::ImageFormat::Png)?)
```

**Comment by ShadowLarkens** [comment]
Hrm. Can someone help me with this? Even after reading through the rust error handling guidelines, I can't figure out how to implement non-panick file read/write, as the image library returns [ImageError](https://docs.rs/image/0.23.10/image/error/enum.ImageError.html) instead of Error and there's no type conversion defined for "ImageError -> Error".

**Comment by ShadowLarkens** [comment]
Well it was an adventure, but with PJB's help ~~definitely just doing it for me with how bad I am at rust~~ it finally returns an error and doesn't blow up rust-g when it fails

**Comment by spookydonut** [comment]
@ShadowLarkens new conflicts

**Comment by ShadowLarkens** [comment]
@spookydonut Resolved, let me know if there's anything I need to change to follow #42

File: src/dmi.rs [hunk 3]
```diff
@@ -63,3 +68,32 @@ fn create_png(path: &str, width: &str, height: &str, data: &str) -> Result<()> {
     let mut writer = encoder.write_header()?;
     Ok(writer.write_image_data(&result)?)
 }
+
+#[derive(Debug)]
+enum ResizePngError {
```
Comment (spookydonut) [inline src/dmi.rs:73]: see `pub enum Error`

File: src/dmi.rs [hunk 4]
```diff
@@ -63,3 +68,32 @@ fn create_png(path: &str, width: &str, height: &str, data: &str) -> Result<()> {
...
+    }
+}
+
+impl From<ImageError> for ResizePngError {
+    fn from(error: ImageError) -> Self {
+        ResizePngError::Image(error)
+    }
+}
+
+fn resize_png(path: &str, width: &str, height: &str) -> std::result::Result<(), ResizePngError> {
```
Comment (spookydonut) [inline src/dmi.rs:90]:
```suggestion
fn resize_png(path: &str, width: &str, height: &str) -> std::result::Result<(), Error> {
```

**Review by spookydonut — CHANGES_REQUESTED** [review]

**Comment by spookydonut** [comment]
@aspenluxxxy

File: src/dmi.rs [hunk 5]
```diff
@@ -10,7 +11,11 @@ byond_fn! { dmi_strip_metadata(path) {
 } }
 
 byond_fn! { dmi_create_png(path, width, height, data) {
-    create_png(path, width, height , data).err()
+    create_png(path, width, height, data).err()
+} }
+
+byond_fn! { dmi_resize_png(path, width, height) {
+    resize_png(path, width, height).err().map(|e| format!("{:?}", e))
```
Comment (aspenluxxxy) [inline src/dmi.rs:18]: No, using `rust_g::Error` will handle this

File: src/dmi.rs (hunk 4 above)
Comment (aspenluxxxy) [inline src/dmi.rs:90]:
```suggestion
fn resize_png<P: AsRef<Path>>(path: P, width: u32, height: u32) -> std::result::Result<(), Error> {
```

File: src/dmi.rs [hunk 6]
```diff
@@ -63,3 +68,32 @@ fn create_png(path: &str, width: &str, height: &str, data: &str) -> Result<()> {
...
+    }
+}
+
+fn resize_png(path: &str, width: &str, height: &str) -> std::result::Result<(), ResizePngError> {
+    let width = u32::from_str_radix(width, 10)?;
+    let height = u32::from_str_radix(height, 10)?;
+
+    let img = image::open(path)?;
+
+    let newimg = img.resize(width, height, image::imageops::Nearest);
```
Comment (aspenluxxxy) [inline src/dmi.rs:96]: Wouldn't it be better to allow choice of resizing method? Nearest Neighbor isn't always the best.

File: src/dmi.rs [hunk 7]
```diff
@@ -63,3 +68,32 @@ fn create_png(path: &str, width: &str, height: &str, data: &str) -> Result<()> {
...
+
+    Ok(newimg.save_with_format(&path, image::ImageFormat::Png)?)
```
Comment (aspenluxxxy) [inline src/dmi.rs:98]:
```suggestion
    Ok(newimg.save_with_format(path.as_ref(), image::ImageFormat::Png)?)
```

File: src/dmi.rs (hunk 3 above)
Comment (aspenluxxxy) [inline src/dmi.rs:73]: Just add a variant to `rust_g::Error`

**Review by aspenluxxxy — CHANGES_REQUESTED** [review]

File: src/dmi.rs (hunk 4 above)
Comment (ShadowLarkens) [reply src/dmi.rs:90]: Can BYOND call functions with u32 arguments? I thought it explicitly only worked with strings

Comment (aspenluxxxy) [reply src/dmi.rs:90]: Correct, but you'd convert them to u32 in the `byond_fn!`

Comment (ShadowLarkens) [reply src/dmi.rs:90]: Trying to do that results in the `?` no longer working for the ParseIntErrors because byond_fn! doesn't actually return a Result

Comment (ShadowLarkens) [reply src/dmi.rs:90]: That's probably not exactly why it doesn't work, but I have no idea how to make it stop yelling at me with this

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Please fix the merge conflict

**Comment by ShadowLarkens** [comment]
@Cyberboss Done

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**